handshake/handshake.go

```diff
@@ -0,0 +1,225 @@
+	}
+	if n, err := w.Write(pubKey.N.Bytes()); n != len(pubKey.N.Bytes()) || err != nil {
+		return fmt.Errorf("failed to write the pubkey [%d != %d]: %v", n, len(pubKey.N.Bytes()), err)
+	}
+	return nil
+}
+
+func readPubKey(r io.Reader) (rsa.PublicKey, error) {
+	pubKey := rsa.PublicKey{}
+	var E int64
+	if err := binary.Read(r, binary.LittleEndian, &E); err != nil {
+		return pubKey, err
+	}
+	pubKey.E = int(E)
+	var size uint64
+	if err := binary.Read(r, binary.LittleEndian, &size); err != nil {
+		return pubKey, err
+	}
+	pubKeyBytes := make([]byte, size)
+	if n, err := r.Read(pubKeyBytes); uint64(n) != size || err != nil {
+		return pubKey, fmt.Errorf("failed to write the pubkey [%d != %d]: %v", n, size, err)
+	}
+	pubKey.N = new(big.Int).SetBytes(pubKeyBytes)
+	return pubKey, nil
+}
+
+func writeChallenge(w io.Writer, challenge [32]byte, pubKey *rsa.PublicKey) error {
+	encryptedChallenge, err := rsa.EncryptPKCS1v15(rand.Reader, pubKey, challenge[:])
+	if err != nil {
+		return err
+	}
+	if err := binary.Write(w, binary.LittleEndian, uint64(len(encryptedChallenge))); err != nil {
+		return err
+	}
+	if n, err := w.Write(encryptedChallenge); n != len(encryptedChallenge) || err != nil {
+		return fmt.Errorf("failed to write the pubkey [%d != %d]: %v", n, len(encryptedChallenge), err)
+	}
+	return nil
+}
+
+func readChallenge(r io.Reader, privKey *rsa.PrivateKey) ([32]byte, error) {
+	var size uint64
+	if err := binary.Read(r, binary.LittleEndian, &size); err != nil {
+		return [32]byte{}, err
+	}
+	encryptedChallenge := make([]byte, size)
+	if n, err := r.Read(encryptedChallenge); uint64(n) != size || err != nil {
+		return [32]byte{}, fmt.Errorf("failed to write the pubkey [%d != %d]: %v", n, size, err)
+	}
+	challengeBytes, err := rsa.DecryptPKCS1v15(rand.Reader, privKey, encryptedChallenge)
+	if err != nil || len(challengeBytes) != 32 {
+		return [32]byte{}, fmt.Errorf("failed to decrypt the challenge [%d != 32]: %v", len(challengeBytes), err)
+	}
+	challenge := [32]byte{}
+	copy(challenge[:], challengeBytes)
+	return challenge, nil
+}
+
+func (hs *handshaker) sign(r io.Reader, w io.Writer) error {
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+	n := len(data)
+
+	hash := sha3.Sum256(data)
+	sig, err := hs.signerVerifier.Sign(hash[:])
+	if err != nil {
+		return err
+	}
+	if err := binary.Write(w, binary.LittleEndian, uint64(n+65)); err != nil {
+		return err
+	}
+	if wn, err := w.Write(append(data, sig...)); wn != n+65 || err != nil {
+		return fmt.Errorf("failed to add the signature [%d != %d]: %v", wn, n+65, err)
+	}
+	return nil
+}
+
+func (hs *handshaker) verify(r io.Reader, w io.Writer) error {
+	var size uint64
+	if err := binary.Read(r, binary.LittleEndian, &size); err != nil {
+		return err
+	}
+	data := make([]byte, size)
+	n, err := r.Read(data)
+	if err != nil && uint64(n) != size {
+		return fmt.Errorf("failed to read [%d != %d]: %v", uint64(n), size, err)
+	}
+	hash := sha3.Sum256(data[:size-65])
```

Comment:
We need to check that `size` is greater than 65 bytes, otherwise a malicious client/server can cause a panic. Also, what is this magic number "65"? If you are standardising the fact that a signature produced by the `protocol.SignVerifier` must be 65 bytes then this should be very clear in both the comments and the code itself. At the moment, that assumption is _implicit_ and hidden within a private helper function. Ideally, the signature can be an arbitrary length.